Add eslint with pre-commit hook

chmelevskij opened the pull request:
Noticed quite a few comments about spaces and similar formatting related things. Adding eslint with quite loose configuration which can be stricter in the future.

Globals eventually would move to modules.

Warnings should be addressed one by one and converted to errors.

McGiverGim commented:
To me each tool that helps to do the right code, is a good idea. And I think we must have one. My only doubt is if we must use eslint, sonarlint, prettier, etc.

Let's see what @mikeller can say about this. About project management he has more experience than me.

chmelevskij commented:
I originally thought of adding prettier and eslint. But that would have made this PR even bigger.

Prettier and eslint serve slightly different purposes, one is formatter another one is linter. I think they could co-exist with  sonar cloud. Maybe just sonar cloud could be replace with eslint eventually.

General idea is to add some code quality checks before it even hits PR. Reducing the amount of formatting and style comments and allowing maintainers to focus on the logic instead 😊

mikeller commented:
@chmelevskij: SonarCloud incorporates aspects of linting and aspects of formatting at a configurable level, and does so with good and user friendly integration with GitHub - not sure what value you see in adding more tooling?

McGiverGim commented:
The good is to integrate it with the IDE. I've Eclipse and VSCode integrated with the Betaflight Sonarcloud rules, but I don't know if we can generate a "read only" user and publish their token to allow everybody to do that.

mikeller commented:
@McGiverGim: I'd be surprised if more than 10% of the contributors in this project used an IDE that can be integrated with any of these solutions - I think for most contributors GitHub integration is the most important and accessible solution.

chmelevskij commented:
This integrates with pre-commit hook, not just editor, so anyone using git will see the linting BEFORE making the push.

mikeller reviewed (changes requested):
@chmelevskij: I did not see anything adding this as a pre-commit-hook to git - you must have forgotten to add this to the pull request. Can you please fix this?

chmelevskij commented:
I've added it with [husky](https://github.com/typicode/husky) 😉

https://github.com/betaflight/betaflight-configurator/blob/b7121f490e99598670bdb0b0aeaf74a5d40733b3/package.json#L109-L112

mikeller commented:
@chmelevskij:

> I've added it with husky ￼

Doesn't seem to be working - nothing shows up in GitHub.

chmelevskij commented:
@mikeller this won't show up on GitHub. How this works is if you write code, try to commit and it's not passing the linter it won't allow to even make a commit.

So as simple example if I have something like:
```
const GUI = {};

// and later
GUI = {};
```
eslint won't allow committing it

mikeller commented:
@chmelevskij: You seem to be wrong again. I can open a pull request with these changes in the GitHub UI without any problems.

chmelevskij commented:
Just to verify, you pulled the branch, made that change and tried to commit it on your machine?

none of the git hooks will work with GitHub UI. To integrate it with the PR process lint command will need to be added to CI. This can be either travis as it is currently used, or GitHub Actions for example.

freshollie commented:
@mikeller https://pre-commit.com/

McGiverGim commented:
I think we are talking about different things. If I have understood well:
- @mikeller is commenting that this pre-commit changes are only informative, not mandatory. Anyone can continue doing PR with bad code modifying the local repo before commiting. The SonarCloud revision that we have is mandatory and can't be disabled.
- @chmelevskij is commenting that this pre-commit can give useful information to the user when commiting and that helps the user to push a PR more similar to what SonarCloud will complain.

In my opinion, both of you are right, now is only to decide if this tool is helpful or it will produce another kind of problems (maybe the user is not able to understand it, maintenance, etc.) or if we have another alternative.

mikeller commented:
@McGiverGim: Yeah, I am just not sure of the value we get out of adding yet another way to lint / style our code - SonarCloud seems to be working well, and most users seem to be understanding intuitively what they have to do when the SonarCloud scan for a pull request shows issues.
I doubt that adding eslint to travis CI would make it easier for users to discover what they need to fix. And a considerable number of our contributors are using the GitHub web UI, so adding pre-commit hooks to the git client will not change anything for them.

mikeller commented:
What are thoughts on changing this to be working only for rules that are supplementatry to the checking that we already get from SonarCloud, as discussed in https://github.com/betaflight/betaflight-configurator/issues/2191#issuecomment-706660414?

chmelevskij commented:
@mikeller I could defo narrow this down to things which usually get highlighted.

I think 2 simple ones like whitespace at the end of the file, and semicolons would already save a lot of time without interfering with sonar cloud.

McGiverGim commented:
I think starting with something very basic is the way to go. We can add others rules later if we find this useful.
This two are very good candidates.

chmelevskij commented:
Updated the PR with narrowed down rules. So now the only two rules are for the dangling whitespace and new line at the end.  This will be run on staged files and fixed automatically, so hopefully should reduce the number of sonar cloud marking these two things 😉

One caveat though, is that `*.vue` files won't be run through it yet, because it would need `eslint-plugin-vue` which has few other rules. So will look into adding it separately to narrow down scope for this one

mikeller commented on package.json:
```diff
@@ -122,5 +128,13 @@
+    }
+  },
+  "lint-staged": {
+    "*.(js|vue)": "eslint --cache --fix"
```
I am not a big fan of committing automatically generated / modified code to into a repository - in my opinion this is of little value, and contributors should instead own their changes / nip bad formatting at the source by fixing the auto-formatting in their tooling.
So can I propose that, instead of auto-formatting code as part of every commit (which is un-intuitive as well), that we simply err out in this case? Users who want to use auto-formatting can always do so with a manual command invocation.
Also, the linter should also be run as part of the tests, so that bad code will be found as part of CI/CD - as I explained before, a considerable fraction of contributors are not running a full local build chain, and will not be able to use this locally.

chmelevskij replied:
What's the preference for CI then? Travis or GitHub action?

mikeller replied:
Travis isn't working for us any more, since they have introduced a fixed limit on machine hours for open source projects, so it is Azure (currently used) or GitHub Action.
I think the simplest way is to simply add the linter checks to the `test` yarn task, which is already run by CI (and should be run by developers as well to make sure that their changes aren't breaking any tests.

chmelevskij commented:
Removed `lint-staged`, also added `vue` plugin. since otherwise linter doesn't work on vue files.

mikeller reviewed (approved):
Looking good now - we can add this to the test run in a later pull request.
